File: pkg/adminui/harpoon.go
package adminui

import (
	"net/http"
	"strings"
	"time"

	"go.openai.org/api/tunnel-client/pkg/config"
	"go.openai.org/api/tunnel-client/pkg/harpoon"
)

type harpoonStatusResponse struct {
	Enabled         bool   `json:"enabled"`
	Reason          string `json:"reason,omitempty"`
	CapturePayloads bool   `json:"capture_payloads"`
}

type harpoonTargetsResponse struct {
	Targets []harpoonTargetResponse `json:"targets"`
}

type harpoonTargetResponse struct {
	Label              string `json:"label"`
	URL                string `json:"url"`
	Description        string `json:"description,omitempty"`
	AllowPlaintextHTTP bool   `json:"allow_plaintext_http"`
	MaxResponseBytes   int    `json:"max_response_bytes"`
	MaxRedirects       int    `json:"max_redirects"`
}

type harpoonCallsResponse struct {
	Calls []harpoonCallResponse `json:"calls"`
}

type harpoonCallResponse struct {
	Timestamp    time.Time `json:"timestamp"`
	Label        string    `json:"label"`
	URL          string    `json:"url"`
	Method       string    `json:"method"`
	Status       int       `json:"status"`
	LatencyMS    int       `json:"latency_ms"`
	ReqBytes     int       `json:"req_bytes"`
	RespBytes    int       `json:"resp_bytes"`
	Error        string    `json:"error,omitempty"`
	RequestBody  *string   `json:"request_body,omitempty"`
	ResponseBody *string   `json:"response_body,omitempty"`
	BodyIsBase64 *bool     `json:"body_is_base64,omitempty"`
}

func handleHarpoonStatus(registry *harpoon.Registry, cfg *config.HarpoonConfig) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		writeJSON(w, http.StatusOK, buildHarpoonStatus(registry, cfg))
	}
}

func handleHarpoonTargets(registry *harpoon.Registry, cfg *config.HarpoonConfig) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		writeJSON(w, http.StatusOK, buildHarpoonTargets(registry, cfg))
	}
}

func handleHarpoonCalls(buffer *harpoon.CallBuffer, cfg *config.HarpoonConfig) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		label := ""
		if r != nil && r.URL != nil {
			label = strings.TrimSpace(r.URL.Query().Get("label"))
		}
		limit := parseLimit(r, 10, 10)
		writeJSON(w, http.StatusOK, buildHarpoonCalls(buffer, cfg, label, limit))
	}
}

func buildHarpoonStatus(registry *harpoon.Registry, cfg *config.HarpoonConfig) harpoonStatusResponse {
	enabled := false
	reason := ""
	if registry == nil {
		reason = "harpoon not initialized"
	} else if len(registry.Targets()) == 0 {
		reason = "no targets configured"
	} else {
		enabled = true
	}

	capture := false
	if cfg != nil {
		capture = cfg.CapturePayloads
	}

	return harpoonStatusResponse{
		Enabled:         enabled,
		Reason:          reason,
		CapturePayloads: capture,
	}
}

func buildHarpoonTargets(registry *harpoon.Registry, cfg *config.HarpoonConfig) harpoonTargetsResponse {
	if registry == nil {
		return harpoonTargetsResponse{}
	}
	targets := registry.Targets()
	out := make([]harpoonTargetResponse, 0, len(targets))
	allowPlaintext := false
	maxResponse := 0
	maxRedirects := 0
	if cfg != nil {
		allowPlaintext = cfg.AllowPlaintextHTTP
		maxResponse = cfg.MaxResponseBytes
		maxRedirects = cfg.MaxRedirects
	}
	for _, target := range targets {
		url := ""
		if target.BaseURL != nil {
			url = target.BaseURL.String()
		}
		out = append(out, harpoonTargetResponse{
			Label:              target.Label,
			URL:                url,
			Description:        target.Description,
			AllowPlaintextHTTP: allowPlaintext,
			MaxResponseBytes:   maxResponse,
			MaxRedirects:       maxRedirects,
		})
	}
	return harpoonTargetsResponse{Targets: out}
}

func buildHarpoonCalls(buffer *harpoon.CallBuffer, cfg *config.HarpoonConfig, label string, limit int) harpoonCallsResponse {
	if buffer == nil {
		return harpoonCallsResponse{}
	}
	capture := cfg != nil && cfg.CapturePayloads
	entries := buffer.Snapshot(limit, label)
	out := make([]harpoonCallResponse, 0, len(entries))
	for _, entry := range entries {
		call := harpoonCallResponse{
			Timestamp: entry.Timestamp,
			Label:     entry.Label,
			URL:       entry.URL,
			Method:    entry.Method,
			Status:    entry.Status,
			LatencyMS: entry.LatencyMS,
			ReqBytes:  entry.ReqBytes,
			RespBytes: entry.RespBytes,
			Error:     entry.Error,
		}
		if capture {
			req := entry.RequestBody
			resp := entry.ResponseBody
			base64Flag := entry.BodyIsBase64
			call.RequestBody = &req
			call.ResponseBody = &resp
			call.BodyIsBase64 = &base64Flag
		}
		out = append(out, call)
	}
	return harpoonCallsResponse{Calls: out}
}

File: pkg/adminui/assets/harpoon.js
(function () {
  const adminUI = (window.adminUI = window.adminUI || {});
  const $ = adminUI.$ || ((id) => document.getElementById(id));

  function ensurePanel() {
    const mount = $("harpoon-mount");
    const template = document.getElementById("harpoon-template");
    if (!mount || !template || $("panel-harpoon")) return;
    const clone = document.importNode(template.content, true);
    mount.appendChild(clone);
  }

  ensurePanel();

  const statusBadge = $("harpoonStatusBadge");
  const captureBadge = $("harpoonCaptureBadge");
  const disabledEl = $("harpoonDisabled");
  const contentEl = $("harpoonContent");
  const targetsEl = $("harpoonTargets");
  const callsEl = $("harpoonCalls");
  const labelFilterEl = $("harpoonLabelFilter");
  const errEl = $("harpoonErr");
  const refreshBtn = $("harpoonRefresh");

  if (!statusBadge || !captureBadge || !labelFilterEl || !callsEl) return;

  let active = false;
  let poller = null;
  let captureEnabled = false;
  let availableLabels = [];

  function setBadge(el, kind, text) {
    el.className = "badge " + kind;
    el.textContent = text;
  }

  function setDisabledState(disabled, reason) {
    if (disabledEl) {
      disabledEl.style.display = disabled ? "block" : "none";
      disabledEl.textContent = reason || "Harpoon disabled";
    }
    if (contentEl) {
      contentEl.style.display = disabled ? "none" : "grid";
    }
  }

  function formatBytes(bytes) {
    if (bytes == null) return "—";
    return bytes.toString();
  }

  function tryPrettyJSON(text) {
    if (!text) return "";
    try {
      const parsed = JSON.parse(text);
      return JSON.stringify(parsed, null, 2);
    } catch (e) {
      return text;
    }
  }

  function formatPayload(body, isBase64) {
    if (!body) return "—";
    if (isBase64) {
      return "Base64 payload:\n" + body;
    }
    return tryPrettyJSON(body);
  }

  function renderTargets(targets) {
    if (!targetsEl) return;
    targetsEl.textContent = "";
    if (!targets || targets.length === 0) {
      const row = document.createElement("tr");
      const cell = document.createElement("td");
      cell.colSpan = 4;
      cell.className = "muted";
      cell.textContent = "No targets configured.";
      row.appendChild(cell);
      targetsEl.appendChild(row);
      return;
    }

    targets.forEach((t) => {
      const row = document.createElement("tr");
      row.innerHTML =
        "<td class='mono'>" +
        (t.label || "—") +
        "</td>" +
        "<td class='mono'>" +
        (t.url || "—") +
        "</td>" +
        "<td>" +
        (t.description || "—") +
        "</td>" +
        "<td class='mono small'>" +
        "allow_plaintext=" +
        String(!!t.allow_plaintext_http) +
        "<br/>max_response_bytes=" +
        formatBytes(t.max_response_bytes) +
        "<br/>max_redirects=" +
        formatBytes(t.max_redirects) +
        "</td>";
      targetsEl.appendChild(row);
    });
  }

  function renderCalls(calls) {
    callsEl.textContent = "";
    if (!calls || calls.length === 0) {
      const empty = document.createElement("div");
      empty.className = "muted small";
      empty.textContent = "No recent calls.";
      callsEl.appendChild(empty);
      return;
    }

    calls.forEach((call) => {
      const wrapper = document.createElement("div");
      wrapper.className = "harpoon-call";

      const header = document.createElement("div");
      header.className = "harpoon-call-header";

      const toggle = document.createElement("button");
      toggle.className = "harpoon-toggle";
      toggle.textContent = "+";

      const error = call.error || "";
      const statusText = call.status ? call.status.toString() : "—";
      const latency = call.latency_ms != null ? call.latency_ms + "ms" : "—";
      const reqBytes = formatBytes(call.req_bytes);
      const respBytes = formatBytes(call.resp_bytes);

      const timestamp = call.timestamp
        ? new Date(call.timestamp).toISOString()
        : "—";

      header.appendChild(toggle);
      header.appendChild(buildCell(timestamp, "mono small"));
      header.appendChild(buildCell(call.label || "—", "mono small"));
      header.appendChild(buildCell(call.method || "—", "mono small"));
      header.appendChild(buildCell(statusText, "mono small"));
      header.appendChild(buildCell(latency, "mono small"));
      header.appendChild(buildCell(reqBytes + " / " + respBytes, "mono small"));
      header.appendChild(buildCell(error || "—", "small"));

      const details = document.createElement("div");
      details.className = "harpoon-call-details";
      details.style.display = "none";

      if (!captureEnabled) {
        const note = document.createElement("div");
        note.className = "muted small";
        note.textContent = "Payload capture is disabled.";
        details.appendChild(note);
      } else {
        const reqBlock = document.createElement("div");
        reqBlock.className = "harpoon-payload";
        const reqTitle = document.createElement("div");
        reqTitle.className = "muted small";
        reqTitle.textContent = "Request body";
        const reqPre = document.createElement("pre");
        reqPre.className = "pre mono";
        reqPre.textContent = formatPayload(call.request_body, false);
        reqBlock.appendChild(reqTitle);
        reqBlock.appendChild(reqPre);

        const respBlock = document.createElement("div");
        respBlock.className = "harpoon-payload";
        const respTitle = document.createElement("div");
        respTitle.className = "muted small";
        respTitle.textContent = "Response body";
        const respPre = document.createElement("pre");
        respPre.className = "pre mono";
        respPre.textContent = formatPayload(
          call.response_body,
          call.body_is_base64
        );
        respBlock.appendChild(respTitle);
        respBlock.appendChild(respPre);

        details.appendChild(reqBlock);
        details.appendChild(respBlock);
      }

      toggle.addEventListener("click", () => {
        const open = details.style.display === "block";
        details.style.display = open ? "none" : "block";
        toggle.textContent = open ? "+" : "–";
      });

      wrapper.appendChild(header);
      wrapper.appendChild(details);
      callsEl.appendChild(wrapper);
    });
  }

  function buildCell(text, cls) {
    const el = document.createElement("div");
    if (cls) el.className = cls;
    el.textContent = text;
    return el;
  }

  function updateLabelFilter(labels, current) {
    labelFilterEl.textContent = "";
    const all = document.createElement("option");
    all.value = "";
    all.textContent = "all labels";
    labelFilterEl.appendChild(all);
    labels.forEach((label) => {
      const opt = document.createElement("option");
      opt.value = label;
      opt.textContent = label;
      labelFilterEl.appendChild(opt);
    });
    labelFilterEl.value = current || "";
  }

  async function refreshStatus() {
    try {
      const status = await adminUI.fetchJSON("/api/harpoon/status");
      captureEnabled = !!status.capture_payloads;
      setBadge(
        statusBadge,
        status.enabled ? "ok" : "warn",
        "Harpoon: " + (status.enabled ? "enabled" : "disabled")
      );
      setBadge(
        captureBadge,
        captureEnabled ? "warn" : "ok",
        "Capture: " + (captureEnabled ? "on" : "off")
      );
      setDisabledState(!status.enabled, status.reason);
      return status.enabled;
    } catch (e) {
      setBadge(statusBadge, "bad", "Harpoon: error");
      setDisabledState(true, "Harpoon unavailable");
      throw e;
    }
  }

  async function refreshTargets() {
    const data = await adminUI.fetchJSON("/api/harpoon/targets");
    const targets = data.targets || [];
    renderTargets(targets);
    availableLabels = targets.map((t) => t.label).filter(Boolean);
    updateLabelFilter(availableLabels, labelFilterEl.value);
  }

  async function refreshCalls() {
    const label = labelFilterEl.value || "";
    const query = label ? "?label=" + encodeURIComponent(label) : "";
    const data = await adminUI.fetchJSON("/api/harpoon/calls" + query);
    renderCalls(data.calls || []);
  }

  async function refreshAll() {
    errEl.textContent = "";
    try {
      const enabled = await refreshStatus();
      if (!enabled) return;
      await refreshTargets();
      await refreshCalls();
    } catch (e) {
      errEl.textContent = "error: " + e;
    }
  }

  function startPolling() {
    if (poller) return;
    poller = setInterval(() => {
      if (active) refreshAll();
    }, 5000);
  }

  function stopPolling() {
    if (poller) {
      clearInterval(poller);
      poller = null;
    }
  }

  function setActiveTab(name) {
    const next = name === "harpoon";
    if (next === active) return;
    active = next;
    if (active) {
      refreshAll();
      startPolling();
    } else {
      stopPolling();
    }
  }

  labelFilterEl.addEventListener("change", refreshCalls);
  refreshBtn.addEventListener("click", refreshAll);

  if (adminUI.onTabChange) {
    adminUI.onTabChange(setActiveTab);
  }

  const selected = document.querySelector(".tab[aria-selected='true']");
  if (selected && selected.dataset && selected.dataset.tab) {
    setActiveTab(selected.dataset.tab);
  }
})();
